### PR TITLE
Enable Building on OSX

### PR DESCRIPTION
**Pull request by keeeeenw**
- Added Apple specific checks to CMakeLists and source code
- Fixed build issue related to LoopClosing.h KeyFrameAndPose definition
- Removed --lboost_serialization from CMakeLists
- Removed --lcrypto from CMakeLists
- Tested using OSX 10.15.5. Build passes and euroc_examples.sh appears
to be working.

Notes:
I also had issues finding openssl/md5.h header on OSX even after installing openssl using Homebrew. https://superuser.com/questions/1034164/missing-md5-h-header-as-i-install-cmusfm-scrobbler-for-cmus-on-os-x provided detailed solutions.

### CODE REVIEW

**Comment by keeeeenw** [comment]
I forgot to mention that I also had this issue while running euroc_eval_examples.sh.
I had to comment out the block for starting the Viewer on a separate thread https://github.com/UZ-SLAMLab/ORB_SLAM3/blob/master/src/System.cc#L198 in order to resolve this issue.
Apparently, the OSX implementation of OpenCV requires the UI Window to be opened as part of the main thread. Reference: https://github.com/robotology/QA/issues/179
```
2020-07-27 18:47:50.721 stereo_euroc[77581:1981897] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'nextEventMatchingMask should only be called from the Main Thread!'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff3332cbe7 __exceptionPreprocess + 250
	1   libobjc.A.dylib                     0x00007fff6c1045bf objc_exception_throw + 48
	2   AppKit                              0x00007fff30524b65 +[NSEvent _discardTrackingAndCursorEventsIfNeeded] + 0
	3   libpangolin.dylib                   0x0000000108485ab0 +[PangolinNSApplication run_step] + 96
	4   libpangolin.dylib                   0x0000000108482a3e _ZN8pangolin9OsxWindowC2ERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEEiib + 222
	5   libpangolin.dylib                   0x00000001084838ba _ZZN8pangolin24RegisterOsxWindowFactoryEvEN16OsxWindowFactory4OpenERKNS_3UriE + 1002
	6   libpangolin.dylib                   0x000000010840a8c0 _ZN8pangolin15FactoryRegistryINS_14VideoInterfaceEE4OpenERKNS_3UriE + 64
	7   libpangolin.dylib                   0x0000000108409931 _ZN8pangolin19CreateWindowAndBindENSt3__112basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEEiiRKNS_6ParamsE + 1297
	8   libORB_SLAM3.dylib                  0x0000000103cff718 _ZN9ORB_SLAM36Viewer3RunEv + 152
	9   libORB_SLAM3.dylib                  0x0000000103ae6d8e _ZNSt3__114__thread_proxyINS_5tupleIJNS_10unique_ptrINS_15__thread_structENS_14default_deleteIS3_EEEEMN9ORB_SLAM36ViewerEFvvEPS8_EEEEEPvSD_ + 62
	10  libsystem_pthread.dylib             0x00007fff6d4b0109 _pthread_start + 148
	11  libsystem_pthread.dylib             0x00007fff6d4abb8b thread_start + 15
)
libc++abi.dylib: terminating with uncaught exception of type NSException
./euroc_eval_examples.sh: line 15: 77581 Abort trap: 6           ./Examples/Stereo/stereo_euroc ./Vocabulary/ORBvoc.txt ./Examples/Stereo/EuRoC.yaml "$pathDatasetEuroc"/MH_01_easy ./Examples/Stereo/EuRoC_TimeStamps/MH01.txt "$pathDatasetEuroc"/MH02 ./Examples/Stereo/EuRoC_TimeStamps/MH02.txt "$pathDatasetEuroc"/MH03 ./Examples/Stereo/EuRoC_TimeStamps/MH03.txt "$pathDatasetEuroc"/MH04 ./Examples/Stereo/EuRoC_TimeStamps/MH04.txt "$pathDatasetEuroc"/MH05 ./Examples/Stereo/EuRoC_TimeStamps/MH05.txt dataset-MH01_to_MH05_stereo
```

**Comment by gtinchev** [comment]
@keeeeenw have a look at the following PR from ORBSLAM2: https://github.com/raulmur/ORB_SLAM2/pull/179/files
I've tried compiling it on a mac so to spare me carrying my heavy linux machine to another country. I've implemented all of the above to get it working on my mac (main thread for video and also thread sleep for the rest of them instead of usleep). This works and starts both the GUI and the main terminal. However, once running and once keypoints are established it crashes with a bus error right after the keyframe detection:

```
Point distribution in KeyFrame: left-> 391 --- right-> 0
[1]    26619 bus error  ./build/orb_advio/mono_advio ./ORB_SLAM3/Vocabulary/ORBvoc.txt  advio15_mono
```

At this point after 12 hours of debugging, I gave up and took my linux laptop with me. I hope you can find a fix for that.

**Comment by keeeeenw** [comment]
@gtinchev Sorry! I don't have a fix and you seem to be further ahead than I was. I am just using Linux for visualizations for now.

**Comment by match08** [comment]
[Mac support for ORB-SLAM3](https://github.com/UZ-SLAMLab/ORB_SLAM3/pull/131)

**Comment by gtinchev** [comment]
@match08 have you tested this? I'm currently in the final week of writing my thesis and can only test afterwards which would be early Oct. If so and it works - WHOOP WHOOP

**Comment by match08** [comment]
> @match08 have you tested this? I'm currently in the final week of writing my thesis and can only test afterwards which would be early Oct. If so and it works - WHOOP WHOOP

Yes, I'm working on some demos, System.cc file `mptViewer = new thread(&Viewer::Run, mpViewer);` MacOS is not supported yet, I'm updating it. MacOS Must thread in main.

**Comment by gtinchev** [comment]
@match08 I think I solved that by doing:
System.cc:
commeting out `mptViewer = new thread(&Viewer::Run, mpViewer);`:

```
 if(bUseViewer)
    {
        mpViewer = new Viewer(this, mpFrameDrawer,mpMapDrawer,mpTracker,strSettingsFile);
        // mptViewer = new thread(&Viewer::Run, mpViewer);
        mpTracker->SetViewer(mpViewer);
        mpLoopCloser->mpViewer = mpViewer;
        mpViewer->both = mpFrameDrawer->both;
    }
```

adding a new function to run the viewer
```
void System::RunViewer()
{
    mpViewer->Run();
}
```

And in the main.cc file:

```
int main(int argc, char **argv) {
    if(argc < 5)
    {
        cerr << endl << "Usage: ./mono_advio path_to_vocabulary path_to_settings output_trajectory_filename path_to_sequence_folder_1 (path_to_image_folder_2 ... path_to_image_folder_N)" << endl;
        return 1;
    }
    const int num_seq = (argc-4);

    // Create SLAM system. It initializes all system threads and gets ready to process frames.
    ORB_SLAM3::System SLAM(argv[1],argv[2],ORB_SLAM3::System::MONOCULAR, true);

    auto resultFuture = std::async(std::launch::async, processing, argv, &SLAM);
    SLAM.RunViewer();
    return resultFuture.get();
}

int processing(char **argv, ORB_SLAM3::System *slamPtr)
{  
    const int num_seq = 1;
    ORB_SLAM3::System& SLAM = *slamPtr;
....
}
```
I hope that helps.

**Comment by match08** [comment]
> @match08 I think I solved that by doing:
> System.cc:
> commeting out `mptViewer = new thread(&Viewer::Run, mpViewer);`:
> 
> ```
>  if(bUseViewer)
>     {
>         mpViewer = new Viewer(this, mpFrameDrawer,mpMapDrawer,mpTracker,strSettingsFile);
>         // mptViewer = new thread(&Viewer::Run, mpViewer);
>         mpTracker->SetViewer(mpViewer);
>         mpLoopCloser->mpViewer = mpViewer;
>         mpViewer->both = mpFrameDrawer->both;
>     }
> ```
> 
> adding a new function to run the viewer
> 
> ```
> void System::RunViewer()
> {
>     mpViewer->Run();
> }
> ```
> 
> And in the main.cc file:
> 
> ```
> int main(int argc, char **argv) {
>     if(argc < 5)
>     {
>         cerr << endl << "Usage: ./mono_advio path_to_vocabulary path_to_settings output_trajectory_filename path_to_sequence_folder_1 (path_to_image_folder_2 ... path_to_image_folder_N)" << endl;
>         return 1;
>     }
>     const int num_seq = (argc-4);
> 
>     // Create SLAM system. It initializes all system threads and gets ready to process frames.
>     ORB_SLAM3::System SLAM(argv[1],argv[2],ORB_SLAM3::System::MONOCULAR, true);
> 
>     auto resultFuture = std::async(std::launch::async, processing, argv, &SLAM);
>     SLAM.RunViewer();
>     return resultFuture.get();
> }
> 
> int processing(char **argv, ORB_SLAM3::System *slamPtr)
> {  
>     const int num_seq = 1;
>     ORB_SLAM3::System& SLAM = *slamPtr;
> ....
> }
> ```
> 
> I hope that helps.

Thanks, your codes I'm not working. crash.

**Comment by match08** [comment]
> @match08 have you tested this? I'm currently in the final week of writing my thesis and can only test afterwards which would be early Oct. If so and it works - WHOOP WHOOP

Mac support for ORB-SLAM3 Tested: [Viewer and AR Example to MacOS](https://github.com/match08/ORB_SLAM3/commit/404da9ef5f8829a12b627ec164c8bea8f50ffbb5)

**Comment by match08** [comment]
> I forgot to mention that I also had this issue while running euroc_eval_examples.sh.
> I had to comment out the block for starting the Viewer on a separate thread https://github.com/UZ-SLAMLab/ORB_SLAM3/blob/master/src/System.cc#L198 in order to resolve this issue.
> Apparently, the OSX implementation of OpenCV requires the UI Window to be opened as part of the main thread. Reference: [robotology/QA#179](https://github.com/robotology/QA/issues/179)
> 
> ```
> 2020-07-27 18:47:50.721 stereo_euroc[77581:1981897] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'nextEventMatchingMask should only be called from the Main Thread!'
> *** First throw call stack:
> (
> 	0   CoreFoundation                      0x00007fff3332cbe7 __exceptionPreprocess + 250
> 	1   libobjc.A.dylib                     0x00007fff6c1045bf objc_exception_throw + 48
> 	2   AppKit                              0x00007fff30524b65 +[NSEvent _discardTrackingAndCursorEventsIfNeeded] + 0
> 	3   libpangolin.dylib                   0x0000000108485ab0 +[PangolinNSApplication run_step] + 96
> 	4   libpangolin.dylib                   0x0000000108482a3e _ZN8pangolin9OsxWindowC2ERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEEiib + 222
> 	5   libpangolin.dylib                   0x00000001084838ba _ZZN8pangolin24RegisterOsxWindowFactoryEvEN16OsxWindowFactory4OpenERKNS_3UriE + 1002
> 	6   libpangolin.dylib                   0x000000010840a8c0 _ZN8pangolin15FactoryRegistryINS_14VideoInterfaceEE4OpenERKNS_3UriE + 64
> 	7   libpangolin.dylib                   0x0000000108409931 _ZN8pangolin19CreateWindowAndBindENSt3__112basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEEiiRKNS_6ParamsE + 1297
> 	8   libORB_SLAM3.dylib                  0x0000000103cff718 _ZN9ORB_SLAM36Viewer3RunEv + 152
> 	9   libORB_SLAM3.dylib                  0x0000000103ae6d8e _ZNSt3__114__thread_proxyINS_5tupleIJNS_10unique_ptrINS_15__thread_structENS_14default_deleteIS3_EEEEMN9ORB_SLAM36ViewerEFvvEPS8_EEEEEPvSD_ + 62
> 	10  libsystem_pthread.dylib             0x00007fff6d4b0109 _pthread_start + 148
> 	11  libsystem_pthread.dylib             0x00007fff6d4abb8b thread_start + 15
> )
> libc++abi.dylib: terminating with uncaught exception of type NSException
> ./euroc_eval_examples.sh: line 15: 77581 Abort trap: 6           ./Examples/Stereo/stereo_euroc ./Vocabulary/ORBvoc.txt ./Examples/Stereo/EuRoC.yaml "$pathDatasetEuroc"/MH_01_easy ./Examples/Stereo/EuRoC_TimeStamps/MH01.txt "$pathDatasetEuroc"/MH02 ./Examples/Stereo/EuRoC_TimeStamps/MH02.txt "$pathDatasetEuroc"/MH03 ./Examples/Stereo/EuRoC_TimeStamps/MH03.txt "$pathDatasetEuroc"/MH04 ./Examples/Stereo/EuRoC_TimeStamps/MH04.txt "$pathDatasetEuroc"/MH05 ./Examples/Stereo/EuRoC_TimeStamps/MH05.txt dataset-MH01_to_MH05_stereo
> ```
[MacOS working:](https://github.com/match08/ORB_SLAM3/tree/ORB_SLAM3_MAC)

File: Thirdparty/g2o/g2o/core/hyper_graph.h [hunk 1]
```diff
@@ -90,7 +90,11 @@ namespace g2o {
       typedef std::set<Edge*>                           EdgeSet;
       typedef std::set<Vertex*>                         VertexSet;
 
+      #if defined(_MSC_VER) || defined(__APPLE__)
+      typedef std::unordered_map<int, Vertex*>          VertexIDMap;
+      #else
       typedef std::tr1::unordered_map<int, Vertex*>     VertexIDMap;
```
Comment (ykpercy) [inline Thirdparty/g2o/g2o/core/hyper_graph.h:96]:
```suggestion
      typedef std::unordered_map<int, Vertex*>     VertexIDMap;
```

**Comment by ykpercy** [comment]
when i run ./tum_vi_examples.sh, it will raise the error information:
`Failed to load image at: /usr/local/cellar/ORB_SLAM3/Dataset/TUM_VI/dataset-room2_512_16/mav0/cam0/data/1520530731400018325.png`
i don't know the reason

**Comment by Jinxuan1000** [comment]
> [Mac support for ORB-SLAM3](https://github.com/UZ-SLAMLab/ORB_SLAM3/pull/131)
I downloaded the branch ORB_SLAM3_MAC and ran the command:  ./Examples/Monocular/mono_euroc ./Vocabulary/ORBvoc.txt ./Examples/Monocular/EuRoC.yaml "$pathDatasetEuroc"/MH01 ./Examples/Monocular/EuRoC_TimeStamps/MH01.txt dataset-MH01_mono

 but there is no interface display, only two txt files are saved, how can I modify it to have a display interface? The logs are as follows:
image

**Comment by match08** [comment]
run my ar demo
https://github.com/match08/ORB_SLAM3/blob/ORB_SLAM3_MAC/Examples/AR




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "UZ-SLAMLab/ORB_SLAM3"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2021年2月26日(星期五) 上午10:27
收件人:&nbsp;"UZ-SLAMLab/ORB_SLAM3"<ORB_SLAM3@noreply.github.com&gt;;
抄送:&nbsp;"卖火柴"<match_2006@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [UZ-SLAMLab/ORB_SLAM3] Enable Building on OSX (#13)





  
Mac support for ORB-SLAM3
 I downloaded the branch ORB_SLAM3_MAC and ran the command:  ./Examples/Monocular/mono_euroc ./Vocabulary/ORBvoc.txt ./Examples/Monocular/EuRoC.yaml "$pathDatasetEuroc"/MH01 ./Examples/Monocular/EuRoC_TimeStamps/MH01.txt dataset-MH01_mono
  
but there is no interface display, only two txt files are saved, how can I modify it to have a display interface? The logs are as follows:

**Comment by match08** [comment]
> > [Mac support for ORB-SLAM3](https://github.com/UZ-SLAMLab/ORB_SLAM3/pull/131)
> > I downloaded the branch ORB_SLAM3_MAC and ran the command:  ./Examples/Monocular/mono_euroc ./Vocabulary/ORBvoc.txt ./Examples/Monocular/EuRoC.yaml "$pathDatasetEuroc"/MH01 ./Examples/Monocular/EuRoC_TimeStamps/MH01.txt dataset-MH01_mono
> 
> but there is no interface display, only two txt files are saved, how can I modify it to have a display interface? The logs are as follows:
> image
run my  AR demo
https://github.com/match08/ORB_SLAM3/blob/ORB_SLAM3_MAC/Examples/AR

**Comment by boahemaa** [comment]
Hi @match08, I tried building your fork, but I get the error mentioned in #342:

"/home/usr/ORB_SLAM3/Thirdparty/DBoW2/DBoW2/BowVector.h:17:10: fatal error:
boost/serialization/serialization.hpp: No such file or directory
17 | #include <boost/serialization/serialization.hpp>

Can you help out with this? I have boost installed via brew and the mentioned file is in usr/local/include.

**Comment by match08** [comment]
> Hi @match08, I tried building your fork, but I get the error mentioned in #342:
> 
> "/home/usr/ORB_SLAM3/Thirdparty/DBoW2/DBoW2/BowVector.h:17:10: fatal error: boost/serialization/serialization.hpp: No such file or directory 17 | #include <boost/serialization/serialization.hpp>
> 
> Can you help out with this? I have boost installed via brew and the mentioned file is in usr/local/include.

# MAC BOOST: brew info boost
set(BOOST_INCLUDE /usr/local/Cellar/boost/1.76.0/include)
set(BOOST_LIBS /usr/local/Cellar/boost/1.76.0/lib)

include_directories(${OpenCV_INCLUDE_DIRS} ${BOOST_INCLUDE})
add_library(DBoW2 SHARED ${SRCS_DBOW2} ${SRCS_DUTILS})
target_link_libraries(DBoW2 ${OpenCV_LIBS} ${BOOST_LIBS})

**Comment by IamShubhamGupto** [comment]
Hey @match08 I was able to replicate your code on my laptop (m1 Mac Pro). Pretty cool stuff and thank you for contribution! Do you know how I can access the top view of the map being generated? I see it has been removed from the gui but how can  I add it back?

**Pull request closed without merging**